### ISSUE TITLE
Allow for using other Learning Rate Schedulers and Optimizers

### ISSUE THREAD

**Issue by PonteIneptique** [emanjavacas/pie#76]
Hey !
I started reading about some other optimizers, as things went through my news feed (stuff like [this](https://penseeartificielle.fr/meilleur-optimizer-ranger-radam-lookahead-fastai/) or [that](https://medium.com/@lessw/new-deep-learning-optimizer-ranger-synergistic-combination-of-radam-lookahead-for-the-best-of-2dc83f79a48d)).

I ended up trying to implement it in pie but wanted to see first what would be the results. The test were done as follow: same training set (\~500k words), same learning rate, same testing set (\~ 63k tokens), cuda, 10 run per configuration. No optimization were done.

For optimizers, were tested Ranger and Adam. I did not try anything else
For learning rate, were tested ReduceLROnPlaeau, CosineAnnealing, Delayed(CosineAnnealing).
Patience overall is 15 steps with improvement. CosineAnnealing T0 is 40, Delay is 10.

Basically, Ranger does not outperform Adam (maybe with other parameters, who knows, as the beta is different from Adam) but Delay(CosineAnnealing) is reaching same results in 40% less time.

If you are okay, PR will be under way.

Results:

![image](https://user-images.githubusercontent.com/1929830/101276563-2f293900-37ae-11eb-9c01-47d5027239d5.png)

![image](https://user-images.githubusercontent.com/1929830/101276566-32242980-37ae-11eb-8400-5cce94f2373f.png)

![image](https://user-images.githubusercontent.com/1929830/101276570-34868380-37ae-11eb-9bd6-39c60354fa1d.png)

![image](https://user-images.githubusercontent.com/1929830/101276573-37817400-37ae-11eb-80da-fb96d8d58c1e.png)

![image](https://user-images.githubusercontent.com/1929830/101276577-3a7c6480-37ae-11eb-9d8f-1a84aaf98fa3.png)

![image](https://user-images.githubusercontent.com/1929830/101276590-4700bd00-37ae-11eb-881f-bdf463e1a8a0.png)

**Comment by emanjavacas** [comment]
We could include an option to select the lr scheduler. That's easy since it's just swapping the pytorch lr scheduler and adapting the step call. If you have the code around feel free to push a PR and we can see how to include it!

**Comment by PonteIneptique** [comment]
So, small update with my old branch, regarding Flat(Cosine)(Delay=10, CosineTmax=40, patience=11): I can definitely recommend it. On a corpus of 1.5M tokens (3 times the previous one), it's not only faster, it's a also scoring higher with less deviation:

![image](https://user-images.githubusercontent.com/1929830/101592899-50c53300-39ef-11eb-885d-270436a768d8.png)
![image](https://user-images.githubusercontent.com/1929830/101592904-528ef680-39ef-11eb-9dfa-333b3da2bb56.png)
![image](https://user-images.githubusercontent.com/1929830/101592907-54f15080-39ef-11eb-8fd1-dcc2605177d5.png)
![image](https://user-images.githubusercontent.com/1929830/101592910-5753aa80-39ef-11eb-801d-3af9b7e96022.png)
![image](https://user-images.githubusercontent.com/1929830/101592916-59b60480-39ef-11eb-90f0-a76024e0acdf.png)

**Comment by PonteIneptique** [comment]
Hey @emanjavacas :)
I was very bugged by the results on Ranger on the first batch, because I remembered running small trainings and having better results than with Adam. Then I remembered I read that Ranger takes a higher learning rate to start with, and that I did use a higher one for my preliminary tests.
So I did it as well with the LASLA corpus, and I scored better results (note that my Adam LR is fine tuned, after close to 100 run to find the best hyperparams), with a 10x higher LR than my Adam one:

![image](https://user-images.githubusercontent.com/1929830/101881530-d8987200-3b94-11eb-9540-2145527e3692.png)
![image](https://user-images.githubusercontent.com/1929830/101881543-ddf5bc80-3b94-11eb-9a39-b52e7f476dbd.png)

**Comment by PonteIneptique** [comment]
I also found out I am using CosineAnnealing the wrong way, but it still perform better than Adam: instead of using T_max as the cycle for which you'd find a cosine curve of LR, I have been using it as a slope (the LR is badly offset, it should be 10 epochs on the right):
![image](https://user-images.githubusercontent.com/1929830/101881666-067db680-3b95-11eb-81d2-cbfbd740db3e.png)

**Comment by PonteIneptique** [comment]
Coming back with new experiences, regarding Ranger vs Adam. 

I have been playing with single tasks models (which indeed improve when fine tuned correctly), and Ranger clearly yields results that are more stable:

![image](https://user-images.githubusercontent.com/1929830/109528825-5e9dc800-7ab5-11eb-8580-bc09a1edcfb8.png)

The second before last and the second are the same config, just the optimizer is changing (without finetuning optimizer hyperparams)
